beluga/test/beluga/estimation/test_cluster_based_estimation.cpp

```
@@ -0,0 +1,373 @@
+
+  const auto adjacent_grid_cells = {
+      SE2d{SO2d{0.0}, Vector2d{+kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{-kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, +kSpatialHashResolution}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, -kSpatialHashResolution}},
+  };
+
+  // test proper
+
+  // only cells beyond the 10% weight percentile to avoid the messy border
+  // between clusters beneath that threshold
+  const auto ten_percent_threshold = calculate_percentile_weight_threshold(map, 0.15);
+  map_cells_to_clusters(map, spatial_hash_function, adjacent_grid_cells, weight_cap);
+
+  auto cells_above_minimum_threshold_view =
+      coordinates | ranges::views::filter([&](const auto& c) { return std::get<2>(c) >= ten_percent_threshold; });
+
+  auto quadrant_1_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) < 18.0 && std::get<1>(c) < 18.0;
+                         });
+  auto quadrant_2_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) >= 18.0 && std::get<1>(c) < 18.0;
+                         });
+  auto quadrant_3_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) < 18.0 && std::get<1>(c) >= 18.0;
+                         });
+  auto quadrant_4_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) >= 18.0 && std::get<1>(c) >= 18.0;
+                         });
+
+  const auto coord_to_hash = [&](const auto& coords) {
+    const auto& [x, y, w] = coords;
+    const auto state = SE2d{SO2d{0.}, Vector2d{x, y}};
+    return spatial_hash_function(state);
+  };
+
+  const auto hash_to_id = [&](const auto& hash) { return map[hash].cluster_id ? map[hash].cluster_id.value() : 9999; };
```

@glpuga hmm, shouldn't all particles be assigned a cluster? If so, why not simply invoke `map[hash].cluster_id.value()` unconditionally and let the exception propagate if broken?